### ISSUE TITLE
Refactor Middleware in API to its own class

### ISSUE THREAD

**Issue by BenjamenMeyer** [falconry/falcon#837]
This is a break out from https://github.com/falconry/falcon/pull/594

Essentially, I want to be able to unit test the individual parts of the middleware. Current implementation does not make this possible. PR #594 introduced a parallel class to the `srmock` called `smmock` to do this. The objection was that the middleware implementation in `falcon.API` might diverge from what is done for `smmock` and thus there are two places to maintain essentially the same code.

So I am proposing the following:
- [ ] Refactor the middleware functionality out of the `falcon.API` object and into its own object in `falcon.middleware`
- [ ] `falcon.API` uses the new object at run-time to process the middleware
- [ ] the test framework makes it easy to interact with the object to test the individual steps of the middleware during unit tests. This may be achievable by updating the `srmock` to add a few functions that call the new middleware and wrap with the calls to the middleware with the `TypeError` checks expressed in PR #594.

**Comment by BenjamenMeyer** [comment]
I reviewed the code more, and I do think this is still necessary - again, focusing on being able to properly test the middleware for unit tests outside of the API invocation.

I like the idea of your manager object; though in reviewing the code (https://github.com/falconry/falcon/blob/master/falcon/api.py#L178) I'd probably try to move all the loops around the middleware invocation into the manager object (which IIRC was part of the original objection of duplication of in my `smmock` functionality in PR #594).

Ideally, this work could clean up https://github.com/falconry/falcon/blob/master/falcon/api.py#L178 such that the flow would be:
- run request middleware (https://github.com/falconry/falcon/blob/master/falcon/api.py#L212)
- handle the request (https://github.com/falconry/falcon/blob/master/falcon/api.py#L229)
- run resource middleware (https://github.com/falconry/falcon/blob/master/falcon/api.py#L239)
- run response middleware (https://github.com/falconry/falcon/blob/master/falcon/api.py#L258)

Something like:
```python

try:
    middleware_manager.process_request(req, resp)
    responder, params, resource, req.uri_template = self._get_responder(req)
else:
    middleware_manager.process_resource(req, resp, resource, params)
finally:
    middleware_manager.process_response(req, resp, resource, req_succeeded)
```

**Comment by vytas7** [comment]
I'm going to close this proposal since it hasn't gathered much interest over time.
I fully agree with @nZac's request of adding middleware afterwards, however, [`App.add_middleware()`](https://falcon.readthedocs.io/en/stable/api/app.html#falcon.App.add_middleware) has been available for quite some time since 3.0+.
